File: apps/web/lib/grades.ts
// Canonical letter grade — TS mirror of the single A–F band table.
// Python source of truth: packages/agents-py/src/onlinejourno_agents/grades.py
// (and onlinejourno_scoring/grades.py, vendored per product). Pinned by the shared
// cases in grades.test.ts. Everything that turns a 0–100 score into a letter
// (Galley, The Audit, Frontmatter, Newsroom) uses these bands — don't fork them.

// (threshold, letter) high→low. The one source of truth.
export const GRADE_BANDS: ReadonlyArray<readonly [number, string]> = [
  [80, "A"],
  [65, "B"],
  [50, "C"],
  [35, "D"],
];

// 0–100 score → A/B/C/D/F. null/undefined → "F" (no signal). MUST match
// grades.py::letter_grade.
export function letterGrade(score: number | null | undefined): string {
  if (score == null) return "F";
  for (const [threshold, letter] of GRADE_BANDS) {
    if (score >= threshold) return letter;
  }
  return "F";
}

File: packages/scoring-py/src/onlinejourno_scoring/grades.py
"""Canonical letter grade — the SINGLE A–F band table for the whole suite.

Promoted from cwv._letter_grade. Everything that turns a 0–100 score into a
letter (Galley, The Audit, Frontmatter, Newsroom) MUST use this, so the same
score always yields the same grade. Don't fork the bands.
"""
from __future__ import annotations

import json
from pathlib import Path

# (threshold, letter) high→low. bands.json is the one source of truth;
# the literal below is only the fallback if the file is missing/corrupt.
try:
    _raw = json.loads((Path(__file__).with_name("bands.json")).read_text())
    GRADE_BANDS: tuple[tuple[float, str], ...] = tuple(
        sorted(((float(v), str(k)) for k, v in _raw.items()), reverse=True)
    )
except (OSError, ValueError):
    GRADE_BANDS = ((80.0, "A"), (65.0, "B"), (50.0, "C"), (35.0, "D"))


def letter_grade(score: float | None) -> str:
    """0–100 score → A/B/C/D/F. None → 'F' (no signal)."""
    if score is None:
        return "F"
    s = float(score)
    for threshold, letter in GRADE_BANDS:
        if s >= threshold:
            return letter
    return "F"
